Allow changing per_connection_buffer_limit_bytes using contour configuration for clusters

rajatvig opened the pull request:
Fixes https://github.com/projectcontour/contour/issues/5264
Fixes #1408

tsaarni commented:
Looks very good! 

One quick comment that I noticed while quickly browsing the PR: new parameters in `parameters.go` would be nice to have in `site/content/docs/main/configuration.md` ([link]( https://projectcontour.io/docs/1.25/configuration/#configuration-file)).  BTW I see there are some misses from previous PRs and we are alrady getting out of sync. Too many different ways to configure I guess 😅

rajatvig commented:
> Looks very good!
> 
> One quick comment that I noticed while quickly browsing the PR: new parameters in `parameters.go` would be nice to have in `site/content/docs/main/configuration.md` ([link](https://projectcontour.io/docs/1.25/configuration/#configuration-file)). BTW I see there are some misses from previous PRs and we are alrady getting out of sync. Too many different ways to configure I guess 😅

Just updated the docs and also added some docs for the `max-requests-per-connection` which was not documented. 

The difference between the two PRs is that the older one had a default so a lot of tests changed. On the other hand, this one uses configuration to pass and only sets the value when it is set by the operator in the configmap or CRD which results in fewer test changes.

Just realised that the same [field](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto#envoy-v3-api-field-config-listener-v3-listener-per-connection-buffer-limit-bytes) is supposed to set on the listener as well which I will follow up in a different PR.

rajatvig commented:
https://github.com/projectcontour/contour/pull/5513 is the other PR to set the same parameter on the Listener. Have updated links on this to point to the cluster configuration vs the listener one as earlier.

skriss commented on apis/projectcontour/v1alpha1/contourconfig.go:
```diff
@@ -605,6 +605,15 @@ type ClusterParameters struct {
 	// +optional
 	MaxRequestsPerConnection *uint32 `json:"maxRequestsPerConnection,omitempty"`
+
+	// Defines the soft limit on size of the listener’s new connection read and write buffers in bytes.
```
```suggestion
	// Defines the soft limit on size of the cluster's new connection read and write buffers in bytes.
```

skriss commented on pkg/config/parameters.go:
```diff
@@ -420,6 +420,13 @@ type ClusterParameters struct {
 	// +optional
 	MaxRequestsPerConnection *uint32 `yaml:"max-requests-per-connection,omitempty"`
+
+	// Defines the soft limit on size of the listener’s new connection read and write buffers
```
```suggestion
	// Defines the soft limit on size of the cluster's new connection read and write buffers
```

rajatvig replied:
Fixed

skriss reviewed (changes requested):
Thanks @rajatvig! Looks great, just a couple nits

skriss commented on site/content/docs/main/configuration.md:
```diff
@@ -167,9 +167,11 @@ _This is Envoy's default setting value and is not explicitly configured by Conto
+|-----------------------------------|--------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| dns-lookup-family                 | string | auto    | This field specifies the dns-lookup-family to use for upstream requests to externalName type Kubernetes services from an HTTPProxy route. Values are: `auto`, `v4`, `v6`, `all` |
+| max-requests-per-connection       | int    | none*   | This field specifies the maximum requests for upstream connections. If not specified, there is no limit                                                                         |
+| per-connection-buffer-limit-bytes | int    | none*   | This field specifies the soft limit on size of the listener’s new connection read and write buffer. If not specified, Envoy defaults of 1MiB apply                              |
```
```suggestion
| per-connection-buffer-limit-bytes | int    | none*   | This field specifies the soft limit on size of the cluster's new connection read and write buffer. If not specified, Envoy defaults of 1MiB apply                              |
```

skriss commented on site/content/docs/main/configuration.md:
```diff
@@ -376,6 +378,12 @@ data:
+    #   If not specified, there is no limit.
+    #   Setting this parameter to 1 will effectively disable keep alive
+    #   max-requests-per-connection: 0
+    #   the soft limit on size of the listener’s new connection read and write buffers
```
```suggestion
    #   the soft limit on size of the cluster's new connection read and write buffers
```

sunjayBhatia commented on internal/dag/httpproxy_processor.go:
```diff
@@ -100,6 +100,9 @@ type HTTPProxyProcessor struct {
 	MaxRequestsPerConnection *uint32
+
+	// PerConnectionBufferLimitBytes defines the soft limit on size of the listener’s new connection read and write buffers.
+	PerConnectionBufferLimitBytes *uint32
```
I think we need to add this config to the Ingress and Gateway API processors as well to ensure this setting is applied to all generated Clusters

rajatvig replied:
Nice catch. Just updated the PR to include both the processors.

skriss commented on cmd/contour/serve.go:
```diff
@@ -1124,24 +1126,27 @@ func (s *Server) getDAGBuilder(dbc dagBuilderConfig) *dag.Builder {
+			ResponseHeadersPolicy:         &responseHeadersPolicy,
+			ConnectTimeout:                dbc.connectTimeout,
+			GlobalExternalAuthorization:   dbc.globalExternalAuthorizationService,
+			MaxRequestsPerConnection:      dbc.maxRequestsPerConnection,
+			PerConnectionBufferLimitBytes: dbc.perConnectionBufferLimitBytes,
```
Need to set these on the `IngressProcessor` above as well

rajatvig replied:
Fixed

skriss commented on site/content/docs/main/configuration.md:
```diff
@@ -167,9 +167,11 @@ _This is Envoy's default setting value and is not explicitly configured by Conto
+| Field Name                        | Type   | Default | Description                                                                                                                                                                     |
+|-----------------------------------|--------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| dns-lookup-family                 | string | auto    | This field specifies the dns-lookup-family to use for upstream requests to externalName type Kubernetes services from an HTTPProxy route. Values are: `auto`, `v4`, `v6`, `all` |
+| max-requests-per-connection       | int    | none*   | This field specifies the maximum requests for upstream connections. If not specified, there is no limit                                                                         |
+| per-connection-buffer-limit-bytes | int    | none*   | This field specifies the soft limit on size of the cluster’s new connection read and write buffer. If not specified, Envoy defaults of 1MiB apply                               |
```
I think the idea with the `*` next to defaults was to indicate defaults that _Envoy_ applied, rather than Contour -- so in this case, `per-connection-buffer-limit-bytes` should list a default of `1MiB` with a `*`, and then `max-requests-per-connection` does not need a `*` since there's no default.

If we're going to use that, please copy the footnote from `Timeout Configuration` to this section as well. It also looks like the leading `*` in that footnote was dropped at some point - please add it back so it's clear what that note refers to.

skriss reviewed (changes requested):
One more docs nit, otherwise LGTM. I think we can see if there are any opportunities for further refactoring with #5458
